### PR TITLE
Add stable dtype and device PyObject conversions and an is_tensor probe (2.15+)

### PR DESCRIPTION
**Pull request by pggPL**
Follow-up to #183323, extending the Python interop shims with dtype/device conversions and a tensor type-check probe, using the same `PyObjectConversionInterface` vtable mechanism (no new machinery).

## Summary
- Add `torch::stable::{dtype_from_pyobject,dtype_to_pyobject}` (`torch.dtype` <-> `ScalarType`) and `torch::stable::{device_from_pyobject,device_to_pyobject}` (`torch.device` <-> `torch::stable::Device`), gated at 2.14, in `torch/csrc/stable/pyobject.h`
- Add `torch::stable::is_tensor_pyobject`: probes whether a Python object is a `torch.Tensor` (or a subclass). `tensor_from_pyobject` errors on non-tensors, so code that receives arbitrary Python objects (e.g. argument conversion in a Python-registered custom op) needs a way to check first; the non-stable equivalent `THPVariable_Check` requires libtorch_python, and checking from Python via `isinstance` costs a Python round-trip per argument
- Backing C shims (`torch_is_tensor_pyobject`, `torch_dtype_from_pyobject`, `torch_dtype_to_pyobject`, `torch_device_from_pyobject`, `torch_device_to_pyobject`) in `torch/csrc/stable/c/shim.h` + `torch/csrc/shim_common.cpp`; no handle type involved — a dtype crosses the ABI boundary as the `int32_t` code already used by `aoti_torch_dtype_*()`, a device as a `(device_type, device_index)` `int32_t` pair in the `aoti_torch_device_type_*()` encoding; codes are translated through `torch::stable::detail::from`/`to` like the existing `Tensor::scalar_type()`/`device()` accessors
- Implementations in libtorch_python via `THPVariable_Check`, `THPDtype`/`torch::getTHPDtype` and `THPDevice`/`THPDevice_New`; like the tensor shims they require libtorch_python loaded at runtime and the GIL held, and the no-op default errors with a clear message in a libtorch-only process (also for the probe — a silent `false` would mask libtorch_python missing)
- The C shims validate dtype/device codes at the boundary before narrowing (an out-of-range code, e.g. device_index=256, errors instead of silently wrapping through the int8-backed c10 types)
- The header-only helpers translate through the stable enum mappings (`torch::stable::detail::from`/`to`), which cover a subset of ScalarTypes / DeviceTypes; a valid `torch.dtype` / `torch.device` outside that subset errors (documented in `pyobject.h` and the stable ABI note)
- Document the new APIs in the Python interop shims section of the LibTorch stable ABI note

## Test plan
- [x] `python -m pytest test/cpp_extensions/test_libtorch_agnostic.py -k pyobject -v` — parametrized roundtrips (7 dtypes incl. float8; cpu/meta/cuda/cuda:1 devices, incl. an index-less device that must not come back as cuda:0), end-to-end `Tensor::scalar_type()`/`device()` -> Python object, the probe on tensors / a tensor subclass / non-tensors, error cases for non-dtype/non-device inputs
- [x] `test_shim --gtest_filter="TorchPyObjectConversion.*"` — 6 tests: no-op fallback returns `AOTI_TORCH_FAILURE` (not a crash) without libtorch_python, and out-of-range dtype/device codes are rejected with clear errors

---
_This PR was prepared with the help of Claude._

### CODE REVIEW

File: torch/csrc/stable/c/shim_function_versions.txt [hunk 1]
```diff
@@ -132,7 +132,12 @@ torch_library_set_python_module: TORCH_VERSION_2_13_0
 torch_new_generator_handle: TORCH_VERSION_2_13_0
 torch_new_stable_ivalue: TORCH_VERSION_2_13_0
 torch_stream_native_handle: TORCH_VERSION_2_13_0
+torch_device_from_pyobject: TORCH_VERSION_2_14_0
```
Comment (janeyx99) [inline torch/csrc/stable/c/shim_function_versions.txt]: The branch cut was yesterday :c so this would be 2_15 now

File: torch/csrc/PyObjectConversion.cpp [hunk 2]
```diff
@@ -6,23 +6,40 @@ namespace torch::detail {
...
 
-// Error message shared by both no-op entry points.
+// Error message shared by all no-op entry points.
 constexpr const char* kNoImplMsg =
-    "Converting between a Python object and a torch::stable::Tensor requires "
-    "libtorch_python to be loaded (e.g. `import torch` in the running "
-    "process). This process linked only libtorch.";
+    "Converting between a Python object and its torch::stable equivalent "
+    "requires libtorch_python to be loaded (e.g. `import torch` in the "
+    "running process). This process linked only libtorch.";
```
Comment (janeyx99) [inline torch/csrc/PyObjectConversion.cpp]: nit this should be generic enough for is_tensor_pyobject too, right?
```suggestion
    "Using APIs relating to conversion between a Python object and its "
    "torch::stable equivalent requires libtorch_python to be loaded (e.g. "
    "`import torch` in the running process). This process linked only "
    "libtorch.";
```

File: torch/csrc/PyObjectConversion.cpp [hunk 3]
```diff
@@ -6,23 +6,40 @@ namespace torch::detail {
...
 
 // Default implementation used until libtorch_python registers the real one.
 // Mirrors NoopPyInterpreterVTable: calling a method is a hard error rather than
 // silent misbehavior.
 struct NoopPyObjectConversion final : PyObjectConversionInterface {
+  // Errors like the rest of the family (rather than returning false): a silent
+  // false would mask libtorch_python not being loaded.
```
Comment (janeyx99) [inline torch/csrc/PyObjectConversion.cpp]: unnecessary

File: torch/csrc/PyObjectConversion.h [hunk 4]
```diff
@@ -1,14 +1,17 @@
...
 #include <torch/csrc/Export.h>
 
 // Indirection that lets the libtorch-only Python-interop stable shims call into
 // code that only libtorch_python can provide (THPVariable_* &co) without
 // libtorch (or the user extension) linking libtorch_python. Today this backs
-// PyObject <-> Tensor conversion; other conversions between Python objects and
-// libtorch types that need libtorch_python can be added as further methods.
+// PyObject <-> Tensor/dtype/device conversion; other conversions between Python
+// objects and libtorch types that need libtorch_python can be added as further
+// methods.
```
Comment (janeyx99) [inline torch/csrc/PyObjectConversion.h]: we don't need to keep updating this everywhere -- can you manually check the comments to not mention PyObject <-> Tensor/dtype/device conversion everywhere?

File: torch/csrc/PyObjectConversion.h [hunk 5]
```diff
@@ -32,6 +35,11 @@ namespace torch::detail {
 struct TORCH_API PyObjectConversionInterface {
   virtual ~PyObjectConversionInterface() = default;
 
+  // Whether obj is a Python torch.Tensor (or a subclass). A cheap probe for
```
Comment (janeyx99) [inline torch/csrc/PyObjectConversion.h]: cheaper than what?

Comment (pggPL) [reply torch/csrc/PyObjectConversion.h]: removed

File: torch/csrc/PyObjectConversion.h (hunk 4 above)
Comment (pggPL) [reply torch/csrc/PyObjectConversion.h]: done

File: torch/csrc/PyObjectConversion.cpp (hunk 3 above)
Comment (pggPL) [reply torch/csrc/PyObjectConversion.cpp]: removed

File: torch/csrc/PyObjectConversion.cpp (hunk 2 above)
Comment (pggPL) [reply torch/csrc/PyObjectConversion.cpp]: done

File: torch/csrc/stable/c/shim_function_versions.txt (hunk 1 above)
Comment (pggPL) [reply torch/csrc/stable/c/shim_function_versions.txt]: done

**Comment by pggPL** [comment]
@janeyx99

**Comment by pggPL** [comment]
@janeyx99 can you have a look?
